include/dali/core/mm/detail/free_list.h

```diff
@@ -323,6 +349,185 @@ class coalescing_free_list : public best_fit_free_list {
+ * of elements already in the list and k is the number of inserted elements.
+ *
+ * The tree nodes are allocated from a pooling allocator.
+ */
+class free_tree {
+ public:
+  void clear() {
+    by_addr_.clear();
+    by_size_.clear();
+  }
+
+  /**
+   * @brief Obtains a best-fit block from the tree of free blocks.
+   *
+   * The function allocates smallest possible block. If the alignment requirements
+   * are not met, the block pointer is aligned and the size is checked taking into account
+   * any necessary padding.
+   *
+   * If there's padding at the beginning (due to alignment) or at the end (due to the block being
+   * larger), new free blocks are created from the padding regions and stored in the tree.
+   *
+   * @return Aligned pointer to a smallest suitable memory block that can fit the required number
+   *         of bytes with specified alignment or nullptr, if not available.
+   */
+  void *get(size_t size, size_t alignment) {
+    for (auto it = by_size_.lower_bound({ size, nullptr }); it != by_size_.end(); ++it) {
+      size_t block_size = it->first;
+      char *base = it->second;
+      char *aligned = detail::align_ptr(base, alignment);
+      size_t front_padding = aligned - base;
+      assert(static_cast<ptrdiff_t>(front_padding) >= 0);
+      // NOTE: block_size - front_padding >= size  can overflow and fail - meh, unsigned size_t
+      if (block_size >= size + front_padding) {
+        by_size_.erase(it);
+        size_t back_padding = block_size - size - front_padding;
+        assert(static_cast<ptrdiff_t>(back_padding) >= 0);
+        if (front_padding) {
+          by_addr_[base] = front_padding;
+          by_size_.insert({front_padding, base});
+        } else {
+          by_addr_.erase(base);
+        }
+        if (back_padding) {
+          by_addr_.insert({ aligned + size, back_padding });
+          by_size_.insert({ back_padding, aligned + size });
+        }
+        return aligned;
+      }
+    }
+    return nullptr;
+  }
+
+  /**
+   * @brief Places the memory block in the tree, joining it with adjacent blocks, if possible.
+   */
+  void put(void *ptr, size_t size) {
+    char *addr = static_cast<char *>(ptr);
+    if (by_addr_.empty()) {
+      by_addr_.insert({ addr, size });
+      by_size_.insert({ size, addr });
+      return;
+    }
+    auto next = by_addr_.lower_bound(addr);
+    auto prev = next != by_addr_.begin() ? std::prev(next) : next;
+    bool join_prev = prev != next && prev->first + prev->second == addr;
+    bool join_next = next != by_addr_.end() && next->first == addr + size;
+
+    if (join_prev) {
+      if (join_next) {
+        by_size_.erase({ prev->second, prev->first });
+        by_size_.erase({ next->second, next->first });
+        prev->second += size + next->second;
+        by_addr_.erase(next);
+      } else {
+        by_size_.erase({ prev->second, prev->first });
+        prev->second += size;
+      }
```

Comment:
```suggestion
      by_size_.erase({ prev->second, prev->first });
      prev->second += size;
      if (join_next) {
        by_size_.erase({ prev->second, prev->first });
        prev->second += next->second;
        by_addr_.erase(next);
      }
```

Reply:
This won't work. The suggested code tries to dereference an erased iterator.

Reply:
There's a bug in the suggestion, but general idea seems sound.
```suggestion
      by_size_.erase({ prev->second, prev->first });
      prev->second += size;
      if (join_next) {
        by_size_.erase({ next->second, next->first });
        prev->second += next->second;
        by_addr_.erase(next);
      }
```

Reply:
Yep. To much of copy paste.